File: sdk/typescript/src/exceptions.ts
export interface ErrorDetails {
  [key: string]: unknown;
}

export class TokVigilError extends Error {
  public errorCode: string | null;
  public details: ErrorDetails | null;

  constructor(
    message: string,
    errorCode: string | null = null,
    details: ErrorDetails | null = null,
  ) {
    super(message);
    this.name = "TokVigilError";
    this.errorCode = errorCode;
    this.details = details;
  }
}

export class AuthenticationError extends TokVigilError {
  constructor(
    message: string = "Unauthorized",
    errorCode: string | null = null,
    details: ErrorDetails | null = null,
  ) {
    super(message, errorCode, details);
    this.name = "AuthenticationError";
  }
}

export class RateLimitError extends TokVigilError {
  public retryAfter: number | null;

  constructor(
    message: string = "Rate limit exceeded",
    retryAfter: number | null = null,
    errorCode: string | null = null,
    details: ErrorDetails | null = null,
  ) {
    super(message, errorCode, details);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
}

export class ValidationError extends TokVigilError {
  constructor(
    message: string = "Validation error",
    errorCode: string | null = null,
    details: ErrorDetails | null = null,
  ) {
    super(message, errorCode, details);
    this.name = "ValidationError";
  }
}

export class NotFoundError extends TokVigilError {
  constructor(
    message: string = "Resource not found",
    errorCode: string | null = null,
    details: ErrorDetails | null = null,
  ) {
    super(message, errorCode, details);
    this.name = "NotFoundError";
  }
}

export class APIError extends TokVigilError {
  public statusCode: number | null;

  constructor(
    message: string = "API error",
    statusCode: number | null = null,
    errorCode: string | null = null,
    details: ErrorDetails | null = null,
  ) {
    super(message, errorCode, details);
    this.name = "APIError";
    this.statusCode = statusCode;
  }
}

export class ConnectionError extends TokVigilError {
  constructor(
    message: string = "Connection failed",
    errorCode: string | null = null,
    details: ErrorDetails | null = null,
  ) {
    super(message, errorCode, details);
    this.name = "ConnectionError";
  }
}

export class TimeoutError extends TokVigilError {
  constructor(
    message: string = "Request timed out",
    errorCode: string | null = null,
    details: ErrorDetails | null = null,
  ) {
    super(message, errorCode, details);
    this.name = "TimeoutError";
  }
}

File: sdk/python/tokvigil/exceptions.py
class TokVigilError(Exception):
    """Base exception for TokVigil SDK."""
    def __init__(self, message: str, error_code: str = None, details: dict = None):
        self.message = message
        self.error_code = error_code
        self.details = details or {}
        super().__init__(self.message)


class AuthenticationError(TokVigilError):
    """Invalid or missing API key."""
    pass


class RateLimitError(TokVigilError):
    """Rate limit exceeded."""
    def __init__(self, message: str, retry_after: int = None, **kwargs):
        super().__init__(message, **kwargs)
        self.retry_after = retry_after


class ValidationError(TokVigilError):
    """Invalid request parameters."""
    pass


class NotFoundError(TokVigilError):
    """Resource not found."""
    pass


class APIError(TokVigilError):
    """General API error."""
    def __init__(self, message: str, status_code: int = None, **kwargs):
        super().__init__(message, **kwargs)
        self.status_code = status_code


class ConnectionError(TokVigilError):
    """Failed to connect to API."""
    pass


class TimeoutError(TokVigilError):
    """Request timed out."""
    pass